pkg/replay.go

```diff
@@ -20,97 +20,118 @@ package ripley
+}
+
+// Ensures we have handled all HTTP request results before exiting
+var waitGroupResults sync.WaitGroup
+
+func Replay(opts *Options) int {
 	// Default exit code
 	var exitCode int = 0
-	// Ensures we have handled all HTTP request results before exiting
-	var waitGroup sync.WaitGroup
 
 	// Send requests for the HTTP client workers to pick up on this channel
-	requests := make(chan *request)
+	requests := make(chan *Request, opts.NumWorkers)
 	defer close(requests)
 
 	// HTTP client workers will send their results on this channel
-	results := make(chan *Result)
+	results := make(chan *Result, opts.NumWorkers)
 	defer close(results)
 
 	// The pacer controls the rate of replay
-	pacer, err := newPacer(phasesStr)
-
+	pacer, err := newPacer(opts.Pace)
 	if err != nil {
 		panic(err)
 	}
 
 	// Read request JSONL input from STDIN
-	scanner := bufio.NewScanner(os.Stdin)
+	reader := bufio.NewReaderSize(os.Stdin, 1024*1024)
+	scanner := bufio.NewScanner(reader)
 
 	// Start HTTP client goroutine pool
-	startClientWorkers(numWorkers, requests, results, dryRun, timeout)
+	startClientWorkers(opts, requests, results)
 	pacer.start()
 
 	for scanner.Scan() {
-		req, err := unmarshalRequest(scanner.Bytes())
+		if pacer.done {
+			break
+		}
+		waitGroupResults.Add(1)
+
+		b := scanner.Bytes()
+		req, err := unmarshalRequest(&b)
 		if err != nil {
 			exitCode = 126
-			result, _ := json.Marshal(Result{
-				StatusCode: 0,
+			res := &Result{
+				StatusCode: -2,
 				Latency:    0,
 				Request:    req,
 				ErrorMsg:   fmt.Sprintf("%v", err),
-			})
-			fmt.Println(string(result))
+			}
 
-			if strict {
+			if opts.Strict {
 				panic(err)
 			}
+
+			results <- res
 			continue
 		}
 
-		if pacer.done {
-			break
-		}
+		duration := pacer.waitDuration(req.Timestamp)
+		time.Sleep(duration)
 
-		// The pacer decides how long to wait between requests
-		waitDuration := pacer.waitDuration(req.Timestamp)
-		time.Sleep(waitDuration)
 		requests <- req
-		waitGroup.Add(1)
-
-		// Goroutine to handle the  HTTP client result
-		go func() {
-			defer waitGroup.Done()
-
-			result := <-results
-
-			// If there's a panic elsewhere, this channel can return nil
-			if result == nil {
-				return
-			}
+	}
 
-			jsonResult, err := json.Marshal(result)
+	if err := scanner.Err(); err != nil {
+		panic(err)
+	}
 
-			if err != nil {
-				panic(err)
-			}
+	waitGroupResults.Wait()
 
-			if !silent {
-				fmt.Println(string(jsonResult))
-			}
-		}()
+	select {
+	case <-metricsRequestReceived:
+	case <-time.After(time.Duration(2) * time.Second):
 	}
 
-	if scanner.Err() != nil {
-		panic(scanner.Err())
+	if opts.PrintStat {
+		metrics.WritePrometheus(os.Stdout, false)
 	}
 
-	waitGroup.Wait()
+	if opts.NlongestPrint {
+		for longestResultsHeap.Len() > 0 {
+			r := heap.Pop(longestResultsHeap)
+			fmt.Println(r.(*Result).toJson())
+		}
+	}
 
 	return exitCode
 }
+
+func b2s(b []byte) string {
+	return unsafe.String(unsafe.SliceData(b), len(b))
```

Comment:
Why not just `return string(b)`? And we don't need a function for it, we can just inline to `string(b)` in all the places we make this conversion.

Reply:
Directly converting string(bytes) or []byte(str) will bring data duplication and poor performance, so in the pursuit of the ultimate performance scenario, need to use the “hack” way to achieve these two types of conversion. Like[ k8s does](https://github.com/kubernetes/apiserver/blob/master/pkg/authentication/token/cache/cached_token_authenticator.go#L290-L297)

Reply:
Right, it converts a byte to a string without memory allocation.
To be honest, I didn't verify it myself, but I come across this function's implementation in many Golang repos. My findings of the root cause of the issue led to [discussion](https://github.com/golang/go/issues/53003#issuecomment-1140276077).

Reply:
Just tested:

```
go test -bench=.
goos: darwin
goarch: amd64
pkg: github.com/loveholidays/ripley/tmp
cpu: VirtualApple @ 2.50GHz
BenchmarkTestJsonRequestString-10    	20031897	        52.29 ns/op
BenchmarkTestJsonRequestB2S-10       	1000000000	         0.5111 ns/op
PASS
ok  	github.com/loveholidays/ripley/tmp	1.980s
```

Reply:
```
go test -run=. -bench=. -benchtime=5s -count 5 -benchmem -cpuprofile=cpu.out -memprofile=mem.out -trace=trace.out

goos: darwin
goarch: amd64
pkg: github.com/loveholidays/ripley/tmp
cpu: VirtualApple @ 2.50GHz
BenchmarkTestJsonRequestString-10    	79934452	        63.35 ns/op	     288 B/op	       1 allocs/op
BenchmarkTestJsonRequestString-10    	90362182	        63.70 ns/op	     288 B/op	       1 allocs/op
BenchmarkTestJsonRequestString-10    	95226892	        63.04 ns/op	     288 B/op	       1 allocs/op
BenchmarkTestJsonRequestString-10    	89002558	        64.04 ns/op	     288 B/op	       1 allocs/op
BenchmarkTestJsonRequestString-10    	89922666	        64.37 ns/op	     288 B/op	       1 allocs/op
BenchmarkTestJsonRequestB2S-10       	1000000000	         0.5158 ns/op	       0 B/op	       0 allocs/op
BenchmarkTestJsonRequestB2S-10       	1000000000	         0.5165 ns/op	       0 B/op	       0 allocs/op
BenchmarkTestJsonRequestB2S-10       	1000000000	         0.5161 ns/op	       0 B/op	       0 allocs/op
BenchmarkTestJsonRequestB2S-10       	1000000000	         0.5161 ns/op	       0 B/op	       0 allocs/op
BenchmarkTestJsonRequestB2S-10       	1000000000	         0.5148 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/loveholidays/ripley/tmp	36.985s
```
string memory:
<img width="773" alt="Screenshot 2023-03-22 at 21 58 30" src="https://user-images.githubusercontent.com/17274766/227047974-1255f39c-3afb-45cb-951d-d5cf79a4bfbf.png">

b2s memory: **no memory usage**

---

b2s cpu:

<img width="848" alt="Screenshot 2023-03-22 at 22 03 25" src="https://user-images.githubusercontent.com/17274766/227048786-041b3bc5-2100-4b79-9573-eb3a8d4a13f3.png">

string cpu:

<img width="636" alt="Screenshot 2023-03-22 at 22 05 07" src="https://user-images.githubusercontent.com/17274766/227049012-d7a055ef-d3b2-46d2-9c13-fb0b7e8d0ead.png">